### PR TITLE
Refactor/registry

### PR DESCRIPTION
**Pull request by hellais**
This pull request is about refactoring proteus-registry to follow the same directory structure and conventions of proteus-orchestrate.

It significantly reduces the amount of code duplications and leads to a bit cleaner directory structure.

Most of the logic has been left intact, and it's primarily the result of copy-pasting things around.

What I would like from a review is:

* Does this new directory structure and code layout look good to you?
* Upon skimming through the code do you see some obvious mistakes I made?

This depends on #38

### CODE REVIEW

**Comment by bassosimone** [comment]
It depends on #38. I don't fancy much doings reviews of nested PRs :).

**Comment by hellais** [comment]
Merged #38 into master and rebased this into master.

**Comment by bassosimone** [comment]
> Does this new directory structure and code layout look good to you?

`simone -q` says:
> Yes.  

> Upon skimming through the code do you see some obvious mistakes I made?

Is reading.

**Comment by bassosimone** [comment]
LOL, @hellais [this](https://github.com/TheTorProject/proteus/blob/refactor/registry/proteus-registry/registry/handler/admin.go#L63) is clearly `simone -vvv` coding.

File: proteus-registry/registry/handler/registry.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+	"github.com/lib/pq"
+	uuid "github.com/satori/go.uuid"
+	common "github.com/thetorproject/proteus/proteus-common"
+	"golang.org/x/crypto/bcrypt"
+)
+
+var ctx = log.WithFields(log.Fields{
+	"pkg": "handler",
+	"cmd": "proteus-registry",
+})
```
Comment (bassosimone) [inline proteus-registry/registry/handler/registry.go]: A possible little nit is that now this `ctx` is shared by all files in the package and it's not obvious where it lives, but I would say this is really not very important.

File: proteus-registry/registry/handler/registry.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+			req.Token, req.ProbeFamily,
+			req.ProbeID)
+		if err != nil {
+			tx.Rollback()
+			ctx.WithError(err).Error("failed to insert into active probes table, rolling back")
+			return "", err
+		}
+	}
+
+	{
```
Comment (bassosimone) [inline proteus-registry/registry/handler/registry.go]: This was part of the previous code. I actually find a little weird how you are using scopes here to separate different logical blocks without creating context. I'd maybe consider using comments to separate blocks.

Comment (bassosimone) [reply proteus-registry/registry/handler/registry.go]: (Unless you objective is having a fresh `query` variable for each block. But then perhaps maybe I would find it more natural to have smaller functions. Is this usage of the scopes idiomatic go?)

File: proteus-registry/registry/handler/registry.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+	if err := tx.Commit(); err != nil {
+		ctx.WithError(err).Error("failed to commit transaction, rolling back")
+		return "", err
+	}
+
+	return clientID, nil
+}
+
+// UpdateHandler device endpoint for registered probes to update their metadata
```
Comment (bassosimone) [inline proteus-registry/registry/handler/registry.go]: I think this comment is wrong. What is the verb? Device? (Probably it was also wrong before.)

File: proteus-registry/registry/handler/registry.go [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+		ctx.WithError(err).Error("failed to update")
+		c.JSON(http.StatusBadRequest,
+			gin.H{"error": err.Error()})
+		return
+	}
+	c.JSON(http.StatusOK,
+		gin.H{"status": "ok"})
+}
+
+// RegisterHandler this is the public endpoint handler for registering new clients
```
Comment (bassosimone) [inline proteus-registry/registry/handler/registry.go]: I would also update this comment.

File: proteus-registry/registry/handler/registry.go (hunk 1 above)
Comment (hellais) [reply proteus-registry/registry/handler/registry.go]: I think this is however quite an improvement over the previous situation, where we actually had less packages. Now at least we have a split between the `handler` and the `registry` package. So it's a little bit more fine grained.

File: proteus-registry/registry/handler/registry.go (hunk 2 above)
Comment (hellais) [reply proteus-registry/registry/handler/registry.go]: I have found this pattern used in other pieces of go software and I do find that it does overall create a bit more tidy go. Perhaps some of these blocks should be split up into separate functions.

That said, this section has not changed during the refactoring and I will keep this comment in mind for future refactoring challenges.

File: proteus-registry/registry/handler/registry.go (hunk 3 above)
Comment (hellais) [reply proteus-registry/registry/handler/registry.go]: It's a "device" endpoint in the sense that it uses the `DeviceAuthenticator` to authenticate registered devices. How would you improve this docstring?

File: proteus-registry/registry/handler/registry.go (hunk 4 above)
Comment (hellais) [reply proteus-registry/registry/handler/registry.go]: I am also not sure how I to improve this comment.

**Comment by hellais** [comment]
I am going to merge this PR (to continue developing), though if you have suggestions for how to improve the docstrings I can integrate those as part of another PR or directly into my other working branch.

File: proteus-registry/registry/handler/registry.go (hunk 1 above)
Comment (bassosimone) [reply proteus-registry/registry/handler/registry.go]: Sure

File: proteus-registry/registry/handler/registry.go (hunk 2 above)
Comment (bassosimone) [reply proteus-registry/registry/handler/registry.go]: Okay, I didn't know this was common practice in Go. Let's leave it like this then (I'll try to remember about this answer on a cookie, without eating the cookie, and will not ask again :-).

File: proteus-registry/registry/handler/registry.go (hunk 3 above)
Comment (bassosimone) [reply proteus-registry/registry/handler/registry.go]: My point is that godoc should be complete sentence. So a possible complete sentence here could be `UpdateHandler is the device endpoint [...]`

File: proteus-registry/registry/handler/registry.go (hunk 4 above)
Comment (bassosimone) [reply proteus-registry/registry/handler/registry.go]: `RegisterHandler is the public endpoint...`

**Pull request merged**